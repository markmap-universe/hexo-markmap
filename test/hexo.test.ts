import { describe, expect, it } from "vitest"
import { tagRegisterMock, filterRegisterMock, generatorRegisterMock } from "./hexo"
import "@/index"

type Hexo = typeof hexo

type GetRegisterParams<T extends keyof Hexo["extend"]> =
    Hexo["extend"][T] extends { register: any }
    ? Parameters<Hexo["extend"][T]['register']>
    : never

describe("Hexo Markmap Plugin", () => {
    // Extract the registered functions from mocks
    const tagRegister = tagRegisterMock.mock.calls[0][1] as GetRegisterParams<'tag'>[1]
    const generatorRegister = generatorRegisterMock.mock.calls[0][1] as GetRegisterParams<'generator'>[1]
    const filterRegister = filterRegisterMock.mock.calls[0][1] as GetRegisterParams<'filter'>[1]

    it("should return rendered HTML from the tag", () => {
        const content = "# Hello World"
        const height = "500px"

        const result = tagRegister.call({ slug: 'test' }, [height], content)

        expect(result).toMatchInlineSnapshot(`
          "<div class="markmap-wrap" 
                
                style="height: 500px"
              >
                <script type="application/json">{"content":"Hello World","children":[],"payload":{"tag":"h1","lines":"0,1"}}</script>
                <script type="application/json">{}</script>
              </div>"
        `)
    })

    it("should return the correct generator", () => {
        type Generator = {
            path: string,
            data: () => string | Buffer,
        }[]

        const result = generatorRegister.call({ path: () => 'test' }, { data: () => Buffer.from('test data') })

        expect(result).toBeDefined()
        result.forEach((item: Generator[number]) => {
            expect(item).toHaveProperty('path')
            expect(item).toHaveProperty('data')
            expect(item.data).toBeInstanceOf(Function)
        })
    })

    it("should return the correct assets HTML from the filter", () => {
        const callTagOnSamePage = () => {
            const content = `
# Hello World
- Hello
- Katex: $x = {-b \\pm \\sqrt{b^2-4ac} \\over 2a}$ <!-- markmap: fold -->
- \`\`\`markdown
  # Hello World
  \`\`\`
            `
            const height = "500px"
            tagRegister.call({ slug: 'test' }, [height], content)
        }
        // Test merge assets
        callTagOnSamePage()

        const data = {
            content: "",
            slug: "test",
        }

        filterRegister.call({}, data)

        expect(data.content).toMatchSnapshot()
  
    })
})
