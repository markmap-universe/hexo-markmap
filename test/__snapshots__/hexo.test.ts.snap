// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hexo Markmap Plugin > should return the correct assets HTML from the filter 1`] = `
"<link rel="stylesheet" href="https://fastly.jsdelivr.net/npm/markmap-toolbar@0.18.10/dist/style.css"><script src="https://fastly.jsdelivr.net/npm/d3@7"></script><script src="https://fastly.jsdelivr.net/npm/markmap-view@0.18.10"></script><script src="https://fastly.jsdelivr.net/npm/markmap-toolbar@0.18.10"></script><link rel="stylesheet" href="https://fastly.jsdelivr.net/npm/katex@0.16.18/dist/katex.min.css"><link rel="stylesheet" href="https://fastly.jsdelivr.net/npm/@highlightjs/cdn-assets@11.11.1/styles/default.min.css"><script>((getMarkmap) => {
          window.WebFontConfig = {
            custom: {
              families: [
                "KaTeX_AMS",
                "KaTeX_Caligraphic:n4,n7",
                "KaTeX_Fraktur:n4,n7",
                "KaTeX_Main:n4,n7,i4,i7",
                "KaTeX_Math:i4,i7",
                "KaTeX_Script",
                "KaTeX_SansSerif:n4,n7,i4",
                "KaTeX_Size1",
                "KaTeX_Size2",
                "KaTeX_Size3",
                "KaTeX_Size4",
                "KaTeX_Typewriter"
              ]
            },
            active: () => {
              getMarkmap().refreshHook.call();
            }
          };
        })(() => window.markmap)</script><script src="https://fastly.jsdelivr.net/npm/webfontloader@1.6.28/webfontloader.js" defer></script><link rel="stylesheet" href="/css/markmap.css"><script src="/js/markmap.js"></script>"
`;
